Source/WebCore/editing/cocoa/AttributedString.mm

```diff
@@ -85,79 +89,202 @@
+                verticalAlignment = NSTextBlockTopAlignment;
+                break;
+            case TextTableBlockVerticalAlignment::Middle:
+                verticalAlignment = NSTextBlockMiddleAlignment;
+                break;
+            case TextTableBlockVerticalAlignment::Bottom:
+                verticalAlignment = NSTextBlockBottomAlignment;
+                break;
+            case TextTableBlockVerticalAlignment::Baseline:
+                verticalAlignment = NSTextBlockBaselineAlignment;
+                break;
+            }
+            [tableBlock setVerticalAlignment:verticalAlignment];
+
+            const ParagraphStyleCommonTableAttributes& attributes = item;
+            configureNSTextBlockFromParagraphStyleCommonTableAttributes(tableBlock.get(), attributes);
+
+            return tableBlock;
+        });
+    };
+
+    RetainPtr<NSMutableParagraphStyle> mutableStyle = adoptNS([[PlatformNSParagraphStyle defaultParagraphStyle] mutableCopy]);
+    [mutableStyle setDefaultTabInterval:style.defaultTabInterval];
+    [mutableStyle setHyphenationFactor:style.hyphenationFactor];
+    [mutableStyle setFirstLineHeadIndent:style.firstLineHeadIndent];
+    [mutableStyle setHeadIndent:style.headIndent];
+    [mutableStyle setHeaderLevel:style.headerLevel];
+    [mutableStyle setTailIndent:style.tailIndent];
+    [mutableStyle setParagraphSpacing:style.paragraphSpacing];
+
+    NSTextAlignment alignment;
+    switch (style.alignment) {
+    case ParagraphStyleAlignment::Left:
+        alignment = NSTextAlignmentLeft;
+        break;
+    case ParagraphStyleAlignment::Right:
+        alignment = NSTextAlignmentRight;
+        break;
+    case ParagraphStyleAlignment::Centre:
+        alignment = NSTextAlignmentCenter;
+        break;
+    case ParagraphStyleAlignment::Justified:
+        alignment = NSTextAlignmentJustified;
+        break;
+    case ParagraphStyleAlignment::Natural:
+        alignment = NSTextAlignmentNatural;
+            break;
+    };
+    [mutableStyle setAlignment:alignment];
+
+    NSWritingDirection writingDirection;
+    switch (style.writingDirection) {
+    case ParagraphStyleWritingDirection::LeftToRight:
+        writingDirection = NSWritingDirectionLeftToRight;
+        break;
+    case ParagraphStyleWritingDirection::RightToLeft:
+        writingDirection = NSWritingDirectionRightToLeft;
+        break;
+    case ParagraphStyleWritingDirection::Natural:
+        writingDirection = NSWritingDirectionNatural;
+        break;
+    }
+    [mutableStyle setBaseWritingDirection:writingDirection];
 
-        if (replacementBlock) {
-            if (!adjustedTextBlocks)
-                adjustedTextBlocks = adoptNS((NSMutableArray<NSTextBlock *> *)[[style textBlocks] mutableCopy]);
-            [adjustedTextBlocks setObject:replacementBlock.get() atIndexedSubscript:index];
+    if (!style.textTableBlockIDs.isEmpty()) {
+        RetainPtr<NSMutableArray> blocks = adoptNS([[NSMutableArray alloc] init]);
```

Comment:
nit: new is a little more brief than alloc and init.  Same a few other places in this PR.
But in this case, initWithCapacity would be more efficient.  Can createNSArray be used for all of these?
Also, the template parameter of the RetainPtr is redundant and unnecessary.